runtime/jcl/common/shared.c

```diff
@@ -546,50 +548,75 @@ getPathProtocolFromURL(JNIEnv* env, jobject url, jmethodID URLgetPathID, jmethod
 	return rc;
 }
 
+static BOOLEAN
+isPathTypeJimage(const char *pathChars, jsize pathLen)
+{
+	char JimageEndsWith[] = DIR_SEPARATOR_STR "lib" DIR_SEPARATOR_STR "modules";
+	IDATA len = LITERAL_STRLEN(JimageEndsWith);
+
+	if (pathLen > len) {
+		const char *endsWith = pathChars + (pathLen - len);
+		if (0 == strncmp(endsWith, JimageEndsWith, len)) {
+			return TRUE;
+		}
+	}
+	return FALSE;
+}
 
 /* THREADING: Can be called multi-threaded */
 static UDATA
-getCpeTypeForProtocol(char* protocol, jsize protocolLen, const char* pathChars, jsize pathLen)
+getCpeTypeForProtocol(JNIEnv *env, const char *protocol, jsize protocolLen, const char *urlPathChars, jsize urlPathLen, char **correctedPathPtr)
 {
+	J9JavaVM *vm = ((J9VMThread *)env)->javaVM;
+
 	Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_Entry();
 
-	if (!protocol) {
-		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_Exit0();
-		return 0;
+	if (NULL == protocol) {
+		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitFail1();
+		return CPE_TYPE_UNKNOWN;
+	}
+	if (!correctURLPath(env, urlPathChars, urlPathLen, correctedPathPtr, &(vm->sharedClassConfig->jclStringFarm))) {
+		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitFail2();
+		return CPE_TYPE_UNKNOWN;
 	}
-	if (strncmp(protocol, "jar", 4)==0) {
+	if (0 == strncmp(protocol, "jar", 4)) {
 		Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitJAR();
 		return CPE_TYPE_JAR;
 	}
-	if (strncmp(protocol, "file", 5)==0) {
-		char* endsWith = (char*)(pathChars + (pathLen-4));
-		if ((strncmp(endsWith, ".jar", 4)==0) || (strncmp(endsWith, ".zip", 4)==0) || strstr(pathChars,"!/") || strstr(pathChars,"!\\")) {
+	if (0 == strncmp(protocol, "file", 5)) {
+		const char *pathChars = *correctedPathPtr;
+		jsize pathLen = (jsize)strlen(pathChars);
```

Comment:
Please declare these at the beginning of the function and assign to them before this `if` statement to avoid calling `strlen()` twice:
```c
	pathChars = *correctedPathPtr;
	pathLen = (jsize)strlen(pathChars);
	if (0 == strncmp(protocol, "file", 5)) {
```
Then line 611 can use those locals:
```c
	Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_UnknownProtocol(protocolLen, protocol, pathLen, pathChars);
```

Reply:
In the event the if case is true line 611 will never be executed. What about an else block at 610 to make it more clear?

Reply:
You're right that `strlen()` only gets called once (I misread the code, thinking it fell out of the `if` for the "file" protocol). I still suggest moving the declarations, so there's only one **_place_** in this function that calls `strlen()` to compute that value.